src/kv/tx.h

```diff
@@ -74,13 +80,101 @@ namespace kv
       return std::make_tuple(typed_view);
     }
 
+    template <class M>
+    std::tuple<typename M::TxView*> get_tuple2(const std::string& map_name)
+    {
+      if (store == nullptr)
+      {
+        throw std::logic_error("New form called on old-style Tx");
+      }
+
+      using MapView = typename M::TxView;
+
+      // If the M is present, its AbstractTxView should be an M::TxView. This
+      // invariant could be broken by set_view_list, which will produce an error
+      // here
+      auto search = view_list.find(map_name);
+      if (search != view_list.end())
+      {
+        auto view = dynamic_cast<MapView*>(search->second.view.get());
+
+        if (view == nullptr)
+        {
+          throw std::logic_error(
+            fmt::format("View over map {} is not of expected type", map_name));
+        }
+
+        return std::make_tuple(view);
+      }
+
+      if (read_version == NoVersion)
+      {
+        // Grab opacity version that all Maps should be queried at.
+        auto txid = store->current_txid();
+        term = txid.term;
+        read_version = txid.version;
+      }
+
+      M* typed_map = nullptr;
+
+      auto type_erased_map = store->get_map(read_version, map_name);
+      if (type_erased_map == nullptr)
+      {
+        // Store doesn't know this map yet - create it dynamically
+        {
+          const auto map_it = created_maps.find(map_name);
+          if (map_it != created_maps.end())
+          {
+            throw std::logic_error("Created map without creating view over it");
+          }
+        }
+        // TODO: Currently assuming all dynamic maps are replicated and private
```

Comment:
This assumption is a big one, and I'm not sure how to deal with it. It seems ugly to have to specify these properties every time you ask for a map, but if we're potentially creating the map we need to know them somehow. Perhaps these properties should be rolled into the map type? We assume in C++ you somewhere have a definition `using MyMap = kv::Map<Foo, Bar>`, and that could easily be `using MyMap = kv::Map<Foo, Bar, PRIVATE, REPLICATED>`. That's really just sugar so you can write `auto view = tx.get<MyMap>("my_map")` rather than `auto view = tx.get<MyMap>("my_map", PRIVATE, REPLICATED)` every time. And we need to make this decision again in `js_generic` - does it create a single type of map, or allow these fields to be set.

Reply:
What about we have `tx.create_view<type>(name, domain, replicated)` (basically the existing `get_view2()`, also returning a view over the new map) and a separate `tx.get_view(name)` which only returns a view? Not saying it's a great idea as this would break the "a map is created the first time a view on it is committed" assumption.

Reply:
Will discuss this in depth later, but the reason I'm wary of adding an explicit `create_view` is that it complicates the programming model. My endpoint implementation just wants to get a view and write to the table - if creation is explicit I either need a _separate_ endpoint that creates the table first (and error handling in my endpoint when the table doesn't exist), or I need a creation path in my endpoint.

As a sketch, I want to do:
```
add_widget = fn(tx, request){
  auto widget_view = tx.get_view<WidgetID, Widget>("widgets");
  widget_view->put(request.widget_id, request.widget);
};
```

and ideally at this level I don't need to know about/care about the table metaproperties (domain, access rights, replicated, etc).

I think separate creation requires something like:
```
create_table = fn(tx, request){
  auto view = tx.create_view(request.table_name, request.table_properties);
  // NB: This creates an empty table?
};

add_widget = fn(tx, request){
  auto widget_view = tx.get_view<WidgetID, Widget>("widgets");
  if (widget_view == nullptr)
    return Error("widgets table does not exist");
  widget_view->put(request.widget_id, request.widget);
};
```

or

```
add_widget = fn(tx, request){
  auto widget_view = tx.get_view<WidgetID, Widget>("widgets");
  if (widget_view == nullptr)
  {
    // In this case, I guess widget_table_properties are specified by the script?
    widget_view = tx.create_view("widgets", widget_table_properties);
  }
  widget_view->put(request.widget_id, request.widget);
};
```

For completeness, another thing we've been considering is something like `"widgets"` is a public unencrypted table, and `"_widgets"` is a private encrypted table. This is neat for now, but doesn't necessarily scale to future table properties. It has the big advantage that its not possible to disagree on these properties in one location from another - they're all included in they key.